### PR TITLE
Issue #1900 Fix removing addon when specifying Required-Vars and Var-Defaults

### LINKED ISSUE #1900: Unable to remove addon even specify Required-Vars and Var-Defaults
### General information

  * Minishift version: `master` branch
  * OS: Linux / macOS / Windows
  * Hypervisor: KVM / xhyve / Hyper-V / VirtualBox

### demoaddon.addon
```
$ cat ~/.minishift/addons/demoaddon/demoaddon.addon
# Name: demoaddon
# Description: Setup and Configure Eclipse Che Template and Image Streams
# Required-Vars: NAMESPACE
# Var-Defaults: NAMESPACE=foo

echo "Installing addon with required var 'NAMESPACE' of value #{NAMESPACE}."
```

### demoaddon.addon.remove
```
$ cat ~/.minishift/addons/demoaddon/demoaddon.addon.remove 
# Name: demoaddon
# Description: some description
# Required-Vars: NAMESPACE
# Var-Defaults: NAMESPACE=foo

echo "Removing addon with required var 'NAMESPACE' of value #{NAMESPACE}."
```

### Expected
Should able to remove addon after specifying `Required-Vars` and `Var-Defaults` values.

### Actual
```
$ minishift addon apply demoaddon
-- Applying addon 'demoaddon':
"Installing addon with required var 'NAMESPACE' of value foo."

$ minishift addon remove demoaddon
-- Removing addon 'demoaddon':Error removing the add-on: The variable(s) NAMESPACE are required by the add-on, but are not defined in the context
```

This issue is not related to https://github.com/minishift/minishift/issues/1698 which is completely different.

### PR DESCRIPTION
**Pull request by budhrg**
Fix #1900 

Critical fix for Che addon.

Now, able to remove addon properly.
```
$ cat ~/.minishift/addons/demoaddon/demoaddon.addon 
# Name: demoaddon
# Description: Demoaddon description
# Url: https://someurl.html
# Required-Vars: FOO
# Var-Defaults: FOO=abc
# OpenShift-Version: >=3.6.0

echo This is demo addon with variable FOO whose value is #{FOO}

$ cat ~/.minishift/addons/demoaddon/demoaddon.addon.remove 
# Name: demoaddon
# Description: Demoaddon description
# Url: https://someurl.html
# Required-Vars: FOO
# Var-Defaults: FOO=abc
# OpenShift-Version: >=3.6.0

echo Removing demo addon with variable FOO whose value is #{FOO}

$ minishift addons install demoaddon
Addon 'demoaddon' installed

$ minishift addons apply demoaddon

This is demo addon with variable FOO whose value is abc

$ minishift addons remove demoaddon
-- Removing addon 'demoaddon':
Removing demo addon with variable FOO whose value is abc
```

### CODE REVIEW

**Comment by budhrg** [comment]
retest this please

**Comment by budhrg** [comment]
retest this please

File: pkg/minishift/addon/manager/addon_manager_test.go [hunk 1]
```diff
@@ -134,12 +134,93 @@ func Test_invalid_addons_get_skipped(t *testing.T) {
...
+	if varDefault != "" {
+		testAddonMap["Var-Defaults"] = fmt.Sprintf("%s", varDefault)
+	}
+	if openshiftVersion != "" {
+		testAddonMap["OpenShift-Version"] = openshiftVersion
+	}
+	return testAddonMap
+}
+
+func getAddOnMeta(testMap map[string]interface{}, t *testing.T) addon.AddOnMeta {
```
Comment (LalatenduMohanty) [inline pkg/minishift/addon/manager/addon_manager_test.go]: I think the function name should be `getAddOnMetadata`

File: pkg/minishift/addon/manager/addon_manager_test.go [hunk 2]
```diff
@@ -134,12 +134,93 @@ func Test_invalid_addons_get_skipped(t *testing.T) {
...
+	addOnMeta := getAddOnMeta(testAddonMap, t)
+	addOn := addon.NewAddOn(addOnMeta, nil, nil, "")
 	context, _ := command.NewExecutionContext(nil, nil)
-	expectedVarName := "foo"
-	varDefaults := []addon.RequiredVar{
-		{Key: expectedVarName, Value: "bar"},
+
+	addNameToContext(addOn, context)
+	if len(context.Vars()) != 1 {
+		t.Fatal(fmt.Sprintf("Expected addon name \"%s\" added to the context.", expectedAddonName))
```
Comment (LalatenduMohanty) [inline pkg/minishift/addon/manager/addon_manager_test.go]: You mean `Add-on name \"%s\" found in the context` right?

File: pkg/minishift/addon/manager/addon_manager_test.go [hunk 3]
```diff
@@ -134,12 +134,93 @@ func Test_invalid_addons_get_skipped(t *testing.T) {
...
+	}
+	context.RemoveFromContext("addon-name")
+	if len(context.Vars()) == 1 {
+		t.Fatal(fmt.Sprintf("Expected addon name \"%s\" removed from the context.", expectedAddonName))
```
Comment (LalatenduMohanty) [inline pkg/minishift/addon/manager/addon_manager_test.go]: s/Expected addon/Add-on/

File: pkg/minishift/addon/manager/addon_manager_test.go (hunk 2 above)
Comment (LalatenduMohanty) [reply pkg/minishift/addon/manager/addon_manager_test.go]: s/addon/add-on/

File: pkg/minishift/addon/manager/addon_manager_test.go [hunk 4]
```diff
@@ -134,12 +134,93 @@ func Test_invalid_addons_get_skipped(t *testing.T) {
...
 	}
-	addVarDefaultsToContext(context, varDefaults)
+}
+
+func TestAddVarDefaultsToContext(t *testing.T) {
+	context, _ := command.NewExecutionContext(nil, nil)
+	expectedVarName := "FOO"
+	expectedVarValue := "foo"
```
Comment (LalatenduMohanty) [inline pkg/minishift/addon/manager/addon_manager_test.go]: s/expectedVarValue/expectedVarDefaultValue/

File: pkg/minishift/addon/manager/addon_manager.go [hunk 5]
```diff
@@ -194,23 +194,20 @@ func (m *AddOnManager) String() string {
 
 func (m *AddOnManager) ApplyAddOn(addOn addon.AddOn, context *command.ExecutionContext) error {
 	fmt.Print(fmt.Sprintf("-- Applying addon '%s':", addOn.MetaData().Name()))
-	context.AddToContext("addon-name", addOn.MetaData().Name())
-	defer context.RemoveFromContext("addon-name")
-
-	err := m.verifyRequiredOpenshiftVersion(context, addOn.MetaData())
-	if err != nil {
+	if err := addNameToContext(addOn, context); err != nil {
```
[the remark this answers is not in the document]
Comment (budhrg) [reply pkg/minishift/addon/manager/addon_manager.go]: Sure, I think also testing it doesn't make much sense. Reverting it.

**Pull request merged**